velox/functions/prestosql/tests/utils/FunctionBaseTest.h

```diff
@@ -250,6 +250,44 @@ class FunctionBaseTest : public testing::Test,
   static std::unordered_set<std::string> getSignatureStrings(
       const std::string& functionName);
 
+  template <typename TInput, typename TOutput>
+  void testConstantEval(
+      const std::optional<TInput>& inputValue,
+      vector_size_t size,
+      const core::CallTypedExprPtr& expr,
+      const std::optional<TOutput>& outputValue) {
+    auto expected = makeConstant<TOutput>(outputValue, size, expr->type());
+    auto input = makeRowVector({makeConstant<TInput>(inputValue, size)});
+    velox::test::assertEqualVectors(expected, evaluate(expr, input));
+  }
+
+  void testDictionary(
```

Comment:
It would be nice to make this method more generic. It could accept any expression (TypedExprPtr vs. CallTypedExprPtr) with any number of arguments (N vs. 1) and it could test both constant and dictionary encodings (testEncodings vs. testDictionary).

The calling code would simply create flat inputs and flat expected results, then call testEncodings(expression, inputs, expected).

One way to think about it would be: given an expression, a list of inputs and expected results, generate all sorts of vectors using these values, evaluate the expression and verify the results. This will make sure that VectorFunctions used in the expression handle encodings properly.

The way it could go would be:

1- wrap all inputs in the same dictionary;
2- wrap each input in its own dictionary;
3- make constants from a few rows of inputs;

When generating dictionaries we need to make sure to cover the following cases:

1- dictionary with and without extra nulls
2- dictionary repeats all values, selects a subset of values, repeats some and ignores other values.

Hope this makes sense.

P.S. Use SCOPED_TRACE to provide debugging info. Use nullsEvery(n) for brevity.

Reply:
@mbasmanova Added `testEncodings` method to cover above cases except the one that I may not understand correctly.